---
title: "Enrichment analysis"
output: html_document
---

Load in libraries and functions to use in later analysis.

```{r, warning=FALSE, message=FALSE}
easypackages::libraries("here","knitr")
code_dir = here("code")
datapath = here("data")
source(file.path(code_dir,"genelistOverlap.R"))

genelistOverlap <- function(list1,list2,backgroundTotal, print_result = TRUE, header = FALSE) {
# genelistOverlap
#
# Calculate enrichment odds ratio and p-value from hypergeometric test to
# answer the question of whether genes from one list are enriched in genes
# from another list.
#
# INPUT
#	list1 or list2 = excel file, tab or comma delimited file with gene IDs
#					 assuming each list has a header.
#	backgroundTotal = specify number of genes in background pool
#
# mvlombardo 21.12.2016

	# Read in libraries and set options
	options(stringsAsFactors = FALSE)
	require(readxl)
	require(tools)
  
  if (is.character(list1)){
    # get the file extension of list1
    ext1 = file_ext(list1)
    
    if (is.element(ext1,c("xlsx","xls","txt","csv"))){
      if (ext1=="xlsx" | ext1=="xls") {
        genes1 = read_excel(list1)
      } else if (ext1=="txt") {
        genes1 = read.delim(list1, header = header)
      } else if (ext1=="csv") {
        genes1 = read.csv(list1, header = header)
      }# if (ext1=="xlsx" | ext1=="xls") {
    } else {
      genes1 = data.frame(list1)
    }# if (is.element(ext1,c("xlsx","xls","txt","csv"))){
  } else if (is.data.frame(list1)){
    genes1 = list1
  }# if (is.character(list1)){
  
  if (is.character(list2)){
    # get the file extension of list1
    ext2 = file_ext(list2)
    
    if (is.element(ext2,c("xlsx","xls","txt","csv"))){
      if (ext2=="xlsx" | ext2=="xls") {
        genes2 = read_excel(list2)
      } else if (ext2=="txt") {
        genes2 = read.delim(list2, header = header)
      } else if (ext1=="csv") {
        genes2 = read.csv(list2, header = header)
      }# if (ext1=="xlsx" | ext1=="xls") {
    } else {
      genes2 = data.frame(list2)
    }# if (is.element(ext1,c("xlsx","xls","txt","csv"))){
  } else if (is.data.frame(list2)){
    genes2 = list2
  }# if (is.character(list2)){
  
	# Find overlapping genes
	gene_mask = is.element(genes1[,1],genes2[,1])
	overlapping_genes = genes1[gene_mask,1]
	gene_overlap = sum(gene_mask)
	ngenes1 = length(genes1[,1])
	ngenes2 = length(genes2[,1])

	# Calculate odds ratio
	A = gene_overlap;
	B = ngenes1-gene_overlap
	if (ngenes2==gene_overlap){
	  # add 0.5 to ngenes2 to avoid OR = Inf
	  C = (ngenes2+0.5)-gene_overlap
	} else {
	  C = ngenes2-gene_overlap
	}
	D = backgroundTotal-C
	OR = (A*D)/(B*C)

	# Calculate p-value from hypergeometric test
	hypergeo_p = sum(dhyper(gene_overlap:ngenes2,ngenes1,backgroundTotal-ngenes1,ngenes2))

	# pack into result
	result = vector(mode = "list", length = 1)
	result[[1]]$list1 = list1
	result[[1]]$list2 = list2
	result[[1]]$backgroundTotal = backgroundTotal
	result[[1]]$OR = OR
	result[[1]]$hypergeo_p = hypergeo_p
	result[[1]]$percent_overlap_list1 = gene_overlap/ngenes1
	result[[1]]$percent_overlap_list2 = gene_overlap/ngenes2
	result[[1]]$gene_overlap = gene_overlap
	result[[1]]$ngenes1 = ngenes1
	result[[1]]$ngenes2 = ngenes2
	result[[1]]$overlapping_genes = overlapping_genes

	# print result to the screen and then return result
	if (print_result){
	  print(sprintf("OR = %f, p = %f",OR,hypergeo_p))
	}
	return(result)
} # function genelistOverlap 
```

Examine how M10-NPC overlaps with co-expression modules from Lombardo et al., (2017). Hierarchical cortical transcriptome disorganization in autism. Molecular Autism, 8(1), 29.

```{r, warning=FALSE, message=FALSE}
backgroundTotal = 8075
pm_mods = c("M1","M2","M3","M4","M5","M6","M7","M8","M9","M10",
            "M11","M12","M13","M14","M15","M16","M17","M18","M19","M20",
            "M21","M22","M23","M24","M25","M26","M27")

cols2use = c("OR","pval")
result = data.frame(matrix(nrow = length(pm_mods), ncol = length(cols2use)))
rownames(result) = pm_mods
colnames(result) = cols2use

M10_NPC = as.character(read.delim(file.path(datapath,"M10_NPC.txt"), header=FALSE)$V1)

for (mod in pm_mods){
  pm_genes = as.character(read.delim(file.path(datapath,sprintf("%s.txt",mod)),header=FALSE)$V1)
  res = genelistOverlap(list1 = M10_NPC, list2 = pm_genes, backgroundTotal =8075, print_result = FALSE)
  result[mod,"OR"] = res[[1]]$OR  
  result[mod,"pval"] = res[[1]]$hypergeo_p
}
result$fdrq = p.adjust(result$pval, method = "fdr")
kable(result)
result["M25",]

# show the overlapping genes between M25 and M10_NPC
M25 = as.character(read.delim(file.path(datapath,"M25.txt"),header=FALSE)$V1)
res = genelistOverlap(list1 = M10_NPC, list2 = M25, backgroundTotal =8075, print_result = FALSE)
res[[1]]$overlapping_genes
```

Examine how the Griesi-Oliveira et al., M10-NPC co-expression module overlaps with the autism-upregulated geneM15 co-expression module from Gandal et al., (2018). Transcriptome-wide isoform-level dysregulation in ASD, schizophrenia, and bipolar disorder. Science, 362(6420), eaat8127.

```{r, warning=FALSE, message=FALSE}
# overlap between M10_NPC and the autism-upregulated geneM15 co-expression module from Gandal et al., 2018, Science
geneM15 = as.character(read.delim(file.path(datapath,"geneM15.txt"), header=FALSE)$V1)
res = genelistOverlap(list1 = M10_NPC, list2 = geneM15, backgroundTotal =8075, print_result = FALSE)
print(sprintf("OR = %s, p = %s",res[[1]]$OR,res[[1]]$hypergeo_p))
res[[1]]$overlapping_genes
```

Overlap between Lombardo et al., M25 co-expression module and the autism-upregulated geneM15 co-expression module from Gandal et al., (2018).

```{r, warning=FALSE, message=FALSE}
# overlap M25 from Lombardo et al., and geneM15 from Gandal et al., 2018, Science
M25 = as.character(read.delim(file.path(datapath,"M25.txt"),header=FALSE)$V1)
res = genelistOverlap(list1 = M25, list2 = geneM15, backgroundTotal =8075, print_result = FALSE)
print(sprintf("OR = %s, p = %s",res[[1]]$OR,res[[1]]$hypergeo_p))
res[[1]]$overlapping_genes
```